src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -2508,10 +2516,20 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         location: Location,
     ) -> ty::InstantiatedPredicates<'tcx> {
         if let Some(closure_region_requirements) = tcx.mir_borrowck(def_id).closure_requirements {
-            let closure_constraints =
-                closure_region_requirements.apply_requirements(tcx, def_id, substs);
+            let closure_constraints = QueryRegionConstraints {
+                outlives: closure_region_requirements.apply_requirements(tcx, def_id, substs),
+
+                // Presently, closures never propagate pick
+                // constraints to their parents -- they are enforced
+                // locally.  This is largely a non-issue as pick
+                // constraints only come from `-> impl Trait` and
+                // friends which don't appear (thus far...) in
```

Comment:
I see no reason why we wouldn't support `|x| -> impl Trait { ... }` one day.
Are we future compatible with such extensions if we do this?

Reply:
So, the TL;DR is that I don't think there's a problem, but let me expand a bit:

First, when analyzing a closure body, all the lifetimes that appear in the `-> impl Trait` are "free lifetimes" from the POV of the closure, so they meet this constraint.

The main challenge will be that these constraints may need to be propagated to the closure's creator, since for closures we effectively infer the set of "where clauses" on their free lifetimes. But there is no reason we can't propagate a "member" constraint and then resolve it in the parent's context.

This *will* however mean that the parent may have a member constraint that refers to other lifetime variables, and not just named lifetimes. But this is also true for `let x: impl Foo<'_>`, so we have to cross that bridge, but that was true anyway.

Interestingly, if I'm not mistaken, much of this gets much easier in the context of polonius, and I don't even think you *need* member constraints as a concept. This is because of the way that polonius tracks *provenance* and not lifetimes -- and it supports "set provenances", like `'a|'b`. This implies that a constraint like "R0 member of ['a, 'b, 'static]" is just `R1 <= {'a, 'b}`. (Let bindings may still be tricky-ish, I've not thought that through yet.)

I think at the end of the day I agree with @cramertj that the language as constructed basically *requires* us to work through these challenges. (Along with a few others.)

Reply:
Cool. :) You may want to add your comment to the code in a processed form now that you've written it. :D